File: android/app/src/main/kotlin/com/example/fit_worker/PoseDataStreamHandler.kt
package com.example.fit_worker

import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import io.flutter.plugin.common.EventChannel


class PoseDataStreamHandler : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null

    // Handle the stream initialization
    override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink
    }

    // Handle the stream cancellation
    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // Method to send results to the Dart side
    fun sendResults(results: PoseLandmarkerResult, inputImageHeight: Int, inputImageWidth: Int) {
        if (results.worldLandmarks().isNotEmpty()) {
            val array: MutableMap<String, Any> = mutableMapOf()
            array["poseResult"] = mapResults(results)
            array["inputImageHeight"] = inputImageHeight
            array["inputImageWidth"] = inputImageWidth

            eventSink?.success(array)
        }
    }

    // Convert the results to a map
    private fun mapResults(results: PoseLandmarkerResult): Map<String, List<Map<String, Any>>> {
        val array : MutableMap<String, List<Map<String, Any>>> = mutableMapOf()

        val worldLandmarks: MutableList<Map<String, Any>> = mutableListOf()
        if (results.worldLandmarks().isNotEmpty()) {
            for (worldLandmark in results.worldLandmarks().first()) {
                val landmarkMap: MutableMap<String, Any> = mutableMapOf()
                landmarkMap["x"] = worldLandmark.x()
                landmarkMap["y"] = worldLandmark.y()
                landmarkMap["z"] = worldLandmark.z()
                if (worldLandmark.visibility().isPresent && worldLandmark.presence().isPresent) {
                    landmarkMap["visibility"] = worldLandmark.visibility().get()
                    landmarkMap["presence"] = worldLandmark.presence().get()
                } else {
                    landmarkMap["visibility"] = 0.0
                    landmarkMap["presence"] = 0.0
                }
                worldLandmarks += landmarkMap
            }
        }

        array["worldLandmarks"] = worldLandmarks

        val normalizedLandmarks: MutableList<Map<String, Any>> = mutableListOf()
        if (results.landmarks().isNotEmpty()) {
            for (normalizedLandmark in results.landmarks().first()) {
                val landmarkMap: MutableMap<String, Any> = mutableMapOf()
                landmarkMap["x"] = normalizedLandmark.x()
                landmarkMap["y"] = normalizedLandmark.y()
                landmarkMap["z"] = normalizedLandmark.z()
                if (normalizedLandmark.visibility().isPresent && normalizedLandmark.presence().isPresent) {
                    landmarkMap["visibility"] = normalizedLandmark.visibility().get()
                    landmarkMap["presence"] = normalizedLandmark.presence().get()
                } else {
                    landmarkMap["visibility"] = 0.0
                    landmarkMap["presence"] = 0.0
                }

                normalizedLandmarks += landmarkMap
            }
        }
        array["normalizedLandmarks"] = normalizedLandmarks
        
        return array
    }

}

File: ios/Runner/PoseDataStreamHandler.swift
//
//  PoseDataStreamHandler.swift
//  Runner
//
//  Created by ios on 06/12/2023.
//

import Foundation
import PoseLandmarkerSDK

class PoseDataStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    func sendResults(results: PoseLandmarkerResult, inputImageHeight: Int, inputImageWidth: Int) {
        if (results.worldLandmarks.isEmpty) {
            return
        }

        var array = [String: Any]()

        array["poseResult"] = mapResults(results: results)
        array["inputImageHeight"] = inputImageHeight
        array["inputImageWidth"] = inputImageWidth

        self.eventSink?(array)
    }
    
    private func mapResults(results: PoseLandmarkerResult) -> [String: Any] {
        var array = [String: Any]()

        var worldLandmarks = [[String: Any]]()
        if !results.worldLandmarks.isEmpty {
            for worldLandmark in results.worldLandmarks.first! {
                var landmarkMap = [String: Any]()
                landmarkMap["x"] = worldLandmark.x
                landmarkMap["y"] = worldLandmark.y
                landmarkMap["z"] = worldLandmark.z
                if worldLandmark.visibility != nil && worldLandmark.presence != nil {
                    landmarkMap["visibility"] = worldLandmark.visibility
                    landmarkMap["presence"] = worldLandmark.presence
                } else {
                    landmarkMap["visibility"] = 0.0
                    landmarkMap["presence"] = 0.0
                }
                worldLandmarks.append(landmarkMap)
            }
        }

        array["worldLandmarks"] = worldLandmarks

        var normalizedLandmarks = [[String: Any]]()
        if !results.landmarks.isEmpty {
            for normalizedLandmark in results.landmarks.first! {
                var landmarkMap = [String: Any]()
                landmarkMap["x"] = normalizedLandmark.x
                landmarkMap["y"] = normalizedLandmark.y
                landmarkMap["z"] = normalizedLandmark.z
                if normalizedLandmark.visibility != nil && normalizedLandmark.presence != nil {
                    landmarkMap["visibility"] = normalizedLandmark.visibility
                    landmarkMap["presence"] = normalizedLandmark.presence
                } else {
                    landmarkMap["visibility"] = 0.0
                    landmarkMap["presence"] = 0.0
                }
                normalizedLandmarks.append(landmarkMap)
            }
        }

        array["normalizedLandmarks"] = normalizedLandmarks

        return array
    }
    
}
